### PR TITLE
U-net experimentation and improvement [WIP]

### PR DESCRIPTION
**Pull request by AbigailMcGovern**
Attempting to improve results from U-net. Experimenting with loss functions, training auxiliary tasks, etc.

## Changes
- Added channelwise DICE loss function (didn't get great results)
- Added weighted BCE loss function (looks pretty good)
- Validation during training 
- Optional channel composition for training (train_unet, channels parameter: tuple of int)
  - nth affinity according to '<axis><affinity>' naming convention (e.g., z-1 for z short range affinity)
  - nearness to centroid using euclidian distance ('centreness') or log (natural) of euclidian distance ('centreness-log')
- Loss plots
- Visualisation of all output files for a given training session. 
  - z, y, and x affinities are split into separate channels (nth affinities are channels in axis 1) 
  - centreness and centreness-log are split into their own channels.

## Issues
- Anisotropy makes z channel info more compressed and it seems like predictions are a little bit less spatially accurate

## Plans
- Experiments with different channel compositions
- Experiments with weighted loss
- Napariboard
- Perhaps these should be saved for a future PR?

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -86,32 +92,65 @@ def train_unet(
...
+    device = torch.device(device_name)
+    if validate:
+        no_iter = (epochs * len(xs)) + (epochs * len(v_xs))
+    else:
+        no_iter = epochs * len(xs)
+    print('------------------------------------------------------------')
+    print(f'Training {weights_are} U-net for {epochs} epochs with batch size 1 ')
+    print(f'Device: {device_name}')
+    print('------------------------------------------------------------')
+    with tqdm(total=no_iter, desc='unet training') as progress:
```
Comment (jni) [inline train.py:225]: imho you should look into custom formatting of the tqdm progress bar so you can include the live loss. That'll make the terminal output much nicer. =)

File: unet.py [hunk 2]
```diff
@@ -5,13 +5,34 @@
...
+                               stride=stride, 
+                               padding=padding0
+                               )
+        self.conv1 = nn.Conv3d(
+                               out_channels, 
+                               out_channels, 
+                               kernel_size=kernel_size, 
+                               stride=stride, 
+                               padding=padding1
+                               )
```
Comment (jni) [inline unet.py:58]: So, I don't like black, but I like this less. :joy: Can I suggest either black-style formatting:

```python
self.conv1 = nn.Conv3d(
    out_channels, 
    out_channels, 
    kernel_size=kernel_size, 
    stride=stride, 
    padding=padding1,
)
```

Or my own concoction, which uses two indents for the argument list, to visually distinguish line continuations from actual indented blocks:

```python
self.conv1 = nn.Conv3d(
        out_channels, 
        out_channels, 
        kernel_size=kernel_size, 
        stride=stride, 
        padding=padding1,
        )
```

Note that in both cases the last argument has a comma. Also I'm personally completely undecided about the placement of the closing parenthesis. =P PEP8 does not make a prescription either way.

One good reason to *not* align code to the opening parenthesis is that then you need to change all the lines if you change the name of the call, or the name of the variable you're assigning to — which results in unnecessary noise in a git diff, for example.

**Review by jni — APPROVED** [review]

**Comment by jni** [comment]
Also, now that everything is working, you might want to try out napariboard on this!
